Text generation final touches (#141)
huggingface/optimum-neuron: docs/source/guides/models.mdx

You will also need to pass Neuron specific parameters to configure the export. Each model architecture has its own set of parameters, as detailed in the next paragraphs.

Once your model has been exported, you can save it either on your local or in the [Hugging Face Model Hub](https://hf.co/models):

```python
# Save the neuron model
>>> model.save_pretrained("a_local_path_for_compiled_neuron_model")

# Push the neuron model to HF Hub
>>> model.push_to_hub(  # doctest: +SKIP
...     "a_local_path_for_compiled_neuron_model", repository_id="my-neuron-repo", use_auth_token=True
... )
```

And the next time when you want to run inference, just load your compiled model which will save you the compilation time:

```python
>>> from optimum.neuron import NeuronModelForSequenceClassification
>>> model = NeuronModelForSequenceClassification.from_pretrained("my-neuron-repo")
```

As you see, there is no need to pass the neuron arguments used during the export as they are
saved in a `config.json` file, and will be restored automatically by `NeuronModelForXXX` class.

## Export and inference of Discriminative NLP models

> As explained in the previous paragraph, you will need only a few modifications to your Transformers code to export and run NLP models:

Reviewer: ```suggestion
As explained in the previous section, you will need only few modifications to your Transformers code to export and run NLP models:
```

</Tip>

## Export and inference of Generative NLP models

As explained in the previous paragraph, you will need only a few modifications to your Transformers code to export and run NLP models:

```diff
from transformers import AutoTokenizer
-from transformers import AutoModelForCausalLM
+from optimum.neuron import NeuronModelForCausalLM

# Instantiate and convert to Neuron a PyTorch checkpoint
-model = AutoModelForCausalLM.from_pretrained("gpt2")
+model = NeuronModelForCausalLM.from_pretrained("gpt2", export=True)

tokenizer = AutoTokenizer.from_pretrained("gpt2")
tokenizer.pad_token_id = tokenizer.eos_token_id

tokens = tokenizer("I really wish ", return_tensors="pt")
with torch.inference_mode():
    sample_output = model.generate(
        **tokens,
        do_sample=True,
        min_length=128,
        max_length=256,
        temperature=0.7,
    )
    outputs = [tokenizer.decode(tok) for tok in sample_output]
    print(outputs)
```

### Configuring the export of a generative model

> There are three main parameters that can be passed to the `from_pretrained()` method to configure how a transformers checkpoint is exported to

Reviewer: Would it make sense to put before the load / inference snippet? And probably add the configuration to the snippet? It could be a little bit confusing for new users.

Author: Done

a neuron optimized model:

- `batch_size` is the number of input sequences that the model will accept,
- `num_cores` is the number of neuron cores used when instantiating the model. Each neuron core has 16 Gb of memory, which means that
bigger models need to be split on multiple cores,

> - `auto_cast_type` specifies the format to encode the weights. It can be one of `f32` (`float32`), `f16` (`float16`) or `bf16` (`bfloat16`).

Reviewer: This also can be applied to other model. I just realized that I did not make a detailed doc for compiler's args for other models. I will make a PR for that as well.

As explained before, these parameters can only be configured during export. This means in particular that

> the `batch_size` of the inputs during inference cannot be changed once the model has been exported.

Reviewer: ```suggestion
the `batch_size` of the inputs during inference should be equal to the `batch_size` used for exporting the model.
```

> If not specified, these parameters will take default values.

Reviewer: What are default values?

Author: I will add a link to the reference documentation.
